Creating an Asset with improperly typed properties panics

If I create an `Asset` with poorly typed properties, as in

```typescript
const b = new aws.s3.Bucket("x");
new aws.s3.BucketObject("a", {
    bucket: b,
    source: new pulumi.asset.StringAsset(new Buffer("bad") as any),
});
```

Pulumi panics with `interface conversion: interface {} is map[string]interface {}, not string`:

```
Updating (dev):

     Type                               Name             Status      Info
     pulumi:pulumi:Stack                s3-encypted-dev  running.    (node:92106) [DEP0005] DeprecationWarning: Buffer() is deprecated due to security and usability issues.
 +   └─ pulumi-nodejs:dynamic:Resource  my-datakey       created
panic: interface conversion: interface {} is map[string]interface {}, not string

goroutine 208 [running]:
github.com/pulumi/pulumi/pkg/resource.DeserializeAsset(0xc0016806c0, 0x1f8ab4f, 0x20, 0x1, 0xc0010a0e08)
	/Users/travis/gopath/src/github.com/pulumi/pulumi/pkg/resource/asset.go:255 +0x459
github.com/pulumi/pulumi/pkg/resource/plugin.UnmarshalPropertyValue(0xc001093770, 0xc0001545a0, 0x48, 0x10100000100, 0xc00101da90, 0x0, 0x0)
	/Users/travis/gopath/src/github.com/pulumi/pulumi/pkg/resource/plugin/rpc.go:290 +0x93e
github.com/pulumi/pulumi/pkg/resource/plugin.UnmarshalProperties(0xc0010936e0, 0xc0001545a0, 0x48, 0x10100000100, 0x2, 0xc0008b5de0, 0x1)
	/Users/travis/gopath/src/github.com/pulumi/pulumi/pkg/resource/plugin/rpc.go:211 +0x25f
github.com/pulumi/pulumi/pkg/resource/deploy.(*resmon).RegisterResource(0xc000afd5e0, 0x21f6e60, 0xc001093c20, 0xc00109c120, 0xc000afd5e0, 0x1daf401, 0xc00109a600)
	/Users/travis/gopath/src/github.com/pulumi/pulumi/pkg/resource/deploy/source_eval.go:717 +0x410
github.com/pulumi/pulumi/sdk/proto/go._ResourceMonitor_RegisterResource_Handler.func1(0x21f6e60, 0xc001093c20, 0x1f3b540, 0xc00109c120, 0x1f05500, 0x2d4e9a0, 0x21f6e60, 0xc001093c20)
	/Users/travis/gopath/src/github.com/pulumi/pulumi/sdk/proto/go/resource.pb.go:823 +0x86
github.com/pulumi/pulumi/vendor/github.com/grpc-ecosystem/grpc-opentracing/go/otgrpc.OpenTracingServerInterceptor.func1(0x21f6e60, 0xc0010936b0, 0x1f3b540, 0xc00109c120, 0xc0005d3040, 0xc0005d3060, 0x0, 0x0, 0x21d1ba0, 0xc0001d6c00)
	/Users/travis/gopath/src/github.com/pulumi/pulumi/vendor/github.com/grpc-ecosystem/grpc-opentracing/go/otgrpc/server.go:61 +0x350
github.com/pulumi/pulumi/sdk/proto/go._ResourceMonitor_RegisterResource_Handler(0x1ebe4e0, 0xc000afd5e0, 0x21f6e60, 0xc0010936b0, 0xc00108a7d0, 0xc000233ac0, 0x21f6e60, 0xc0010936b0, 0xc00133ea80, 0x347)
	/Users/travis/gopath/src/github.com/pulumi/pulumi/sdk/proto/go/resource.pb.go:825 +0x158
github.com/pulumi/pulumi/vendor/google.golang.org/grpc.(*Server).processUnaryRPC(0xc0003a7e00, 0x2206800, 0xc000560780, 0xc001081300, 0xc0010bb0b0, 0x2d1e8e8, 0x0, 0x0, 0x0)
	/Users/travis/gopath/src/github.com/pulumi/pulumi/vendor/google.golang.org/grpc/server.go:972 +0x4b5
github.com/pulumi/pulumi/vendor/google.golang.org/grpc.(*Server).handleStream(0xc0003a7e00, 0x2206800, 0xc000560780, 0xc001081300, 0x0)
	/Users/travis/gopath/src/github.com/pulumi/pulumi/vendor/google.golang.org/grpc/server.go:1252 +0xda6
github.com/pulumi/pulumi/vendor/google.golang.org/grpc.(*Server).serveStreams.func1.1(0xc0009e9e40, 0xc0003a7e00, 0x2206800, 0xc000560780, 0xc001081300)
	/Users/travis/gopath/src/github.com/pulumi/pulumi/vendor/google.golang.org/grpc/server.go:691 +0x9f
created by github.com/pulumi/pulumi/vendor/google.golang.org/grpc.(*Server).serveStreams.func1
	/Users/travis/gopath/src/github.com/pulumi/pulumi/vendor/google.golang.org/grpc/server.go:689 +0xa1
```

--- a/pkg/resource/asset.go
+++ b/pkg/resource/asset.go
@@ -248,19 +248,35 @@ func DeserializeAsset(obj map[string]interface{}) (*Asset, bool, error) {
 	// Else, deserialize the possible fields.
 	var hash string
 	if v, has := obj[AssetHashProperty]; has {
-		hash = v.(string)
+		h, ok := v.(string)
+		if !ok {
+			return &Asset{}, false, errors.Errorf("unexpected asset hash of type %T", v)
+		}
+		hash = h
 	}
 	var text string
 	if v, has := obj[AssetTextProperty]; has {
-		text = v.(string)
+		t, ok := v.(string)
+		if !ok {
+			return &Asset{}, false, errors.Errorf("unexpected asset text of type %T", v)
+		}
+		text = t
 	}
 	var path string
 	if v, has := obj[AssetPathProperty]; has {
-		path = v.(string)
+		p, ok := v.(string)
+		if !ok {
+			return &Asset{}, false, errors.Errorf("unexpected asset path of type %T", v)
+		}
+		path = p
 	}
 	var uri string
 	if v, has := obj[AssetURIProperty]; has {
-		uri = v.(string)
+		u, ok := v.(string)
+		if !ok {
+			return &Asset{}, false, errors.Errorf("unexpected asset URI of type %T", v)
+		}
+		uri = u
 	}
 
 	return &Asset{Hash: hash, Text: text, Path: path, URI: uri}, true, nil
@@ -573,14 +589,23 @@ func DeserializeArchive(obj map[string]interface{}) (*Archive, bool, error) {
 
 	var hash string
 	if v, has := obj[ArchiveHashProperty]; has {
-		hash = v.(string)
+		h, ok := v.(string)
+		if !ok {
+			return &Archive{}, false, errors.Errorf("unexpected archive hash of type %T", v)
+		}
+		hash = h
 	}
 
 	var assets map[string]interface{}
 	if v, has := obj[ArchiveAssetsProperty]; has {
 		assets = make(map[string]interface{})
 		if v != nil {
-			for k, elem := range v.(map[string]interface{}) {
+			m, ok := v.(map[string]interface{})
+			if !ok {
+				return &Archive{}, false, errors.Errorf("unexpected archive contents of type %T", v)
+			}
+
+			for k, elem := range m {
 				switch t := elem.(type) {
 				case *Asset:
 					assets[k] = t
@@ -602,18 +627,26 @@ func DeserializeArchive(obj map[string]interface{}) (*Archive, bool, error) {
 						assets[k] = arch
 					}
 				default:
-					return &Archive{}, false, nil
+					return &Archive{}, false, errors.Errorf("archive member '%v' is not an asset or archive", k)
 				}
 			}
 		}
 	}
 	var path string
 	if v, has := obj[ArchivePathProperty]; has {
-		path = v.(string)
+		p, ok := v.(string)
+		if !ok {
+			return &Archive{}, false, errors.Errorf("unexpected archive path of type %T", v)
+		}
+		path = p
 	}
 	var uri string
 	if v, has := obj[ArchiveURIProperty]; has {
-		uri = v.(string)
+		u, ok := v.(string)
+		if !ok {
+			return &Archive{}, false, errors.Errorf("unexpected archive URI of type %T", v)
+		}
+		uri = u
 	}
 
 	return &Archive{Hash: hash, Assets: assets, Path: path, URI: uri}, true, nil

--- a/pkg/resource/plugin/rpc.go
+++ b/pkg/resource/plugin/rpc.go
@@ -291,6 +291,8 @@ func UnmarshalPropertyValue(v *structpb.Value, opts MarshalOptions) (*resource.P
 			if err != nil {
 				return nil, err
 			}
+			// This can only be false with a non-nil error if there is a signature match. We've already verified the
+			// signature.
 			contract.Assert(isasset)
 			if opts.ComputeAssetHashes {
 				if err = asset.EnsureHash(); err != nil {
@@ -304,6 +306,8 @@ func UnmarshalPropertyValue(v *structpb.Value, opts MarshalOptions) (*resource.P
 			if err != nil {
 				return nil, err
 			}
+			// This can only be false with a non-nil error if there is a signature match. We've already verified the
+			// signature.
 			contract.Assert(isarchive)
 			if opts.ComputeAssetHashes {
 				if err = archive.EnsureHash(); err != nil {

--- a/pkg/resource/plugin/rpc_test.go
+++ b/pkg/resource/plugin/rpc_test.go
@@ -19,10 +19,18 @@ import (
 	"runtime"
 	"testing"
 
+	structpb "github.com/golang/protobuf/ptypes/struct"
 	"github.com/stretchr/testify/assert"
 
 	"github.com/pulumi/pulumi/pkg/resource"
+	"github.com/pulumi/pulumi/pkg/util/contract"
 )
+
+func setProperty(s *structpb.Value, k string, v interface{}) {
+	marshaled, err := MarshalPropertyValue(resource.NewPropertyValue(v), MarshalOptions{})
+	contract.Assert(err == nil)
+	s.GetStructValue().Fields[k] = marshaled
+}
 
 func TestAssetSerialize(t *testing.T) {
 	// Ensure that asset and archive serialization round trips.
@@ -41,6 +49,27 @@ func TestAssetSerialize(t *testing.T) {
 	assert.True(t, assetDes.IsText())
 	assert.Equal(t, text, assetDes.Text)
 	assert.Equal(t, "e34c74529110661faae4e121e57165ff4cb4dbdde1ef9770098aa3695e6b6704", assetDes.Hash)
+
+	// Ensure that an invalid asset produces an error.
+	setProperty(assetProps, resource.AssetHashProperty, 0)
+	_, err = UnmarshalPropertyValue(assetProps, MarshalOptions{})
+	assert.Error(t, err)
+	setProperty(assetProps, resource.AssetHashProperty, asset.Hash)
+
+	setProperty(assetProps, resource.AssetTextProperty, 0)
+	_, err = UnmarshalPropertyValue(assetProps, MarshalOptions{})
+	assert.Error(t, err)
+	setProperty(assetProps, resource.AssetTextProperty, "")
+
+	setProperty(assetProps, resource.AssetPathProperty, 0)
+	_, err = UnmarshalPropertyValue(assetProps, MarshalOptions{})
+	assert.Error(t, err)
+	setProperty(assetProps, resource.AssetPathProperty, "")
+
+	setProperty(assetProps, resource.AssetURIProperty, 0)
+	_, err = UnmarshalPropertyValue(assetProps, MarshalOptions{})
+	assert.Error(t, err)
+	setProperty(assetProps, resource.AssetURIProperty, "")
 
 	arch, err := resource.NewAssetArchive(map[string]interface{}{"foo": asset})
 	assert.Nil(t, err)
@@ -68,6 +97,27 @@ func TestAssetSerialize(t *testing.T) {
 		// Go 1.10 introduced breaking changes to archive/zip and archive/tar headers
 		assert.Equal(t, "27ab4a14a617df10cff3e1cf4e30cf510302afe56bf4cc91f84041c9f7b62fd8", archDes.Hash)
 	}
+
+	// Ensure that an invalid archive produces an error.
+	setProperty(archProps, resource.ArchiveHashProperty, 0)
+	_, err = UnmarshalPropertyValue(archProps, MarshalOptions{})
+	assert.Error(t, err)
+	setProperty(archProps, resource.ArchiveHashProperty, arch.Hash)
+
+	setProperty(archProps, resource.ArchiveAssetsProperty, 0)
+	_, err = UnmarshalPropertyValue(archProps, MarshalOptions{})
+	assert.Error(t, err)
+	setProperty(archProps, resource.ArchiveAssetsProperty, nil)
+
+	setProperty(archProps, resource.ArchivePathProperty, 0)
+	_, err = UnmarshalPropertyValue(archProps, MarshalOptions{})
+	assert.Error(t, err)
+	setProperty(archProps, resource.ArchivePathProperty, "")
+
+	setProperty(archProps, resource.ArchiveURIProperty, 0)
+	_, err = UnmarshalPropertyValue(archProps, MarshalOptions{})
+	assert.Error(t, err)
+	setProperty(archProps, resource.ArchiveURIProperty, "")
 }
 
 func TestComputedSerialize(t *testing.T) {

--- a/sdk/nodejs/tests/runtime/langhost/cases/012.assets_archive/index.js
+++ b/sdk/nodejs/tests/runtime/langhost/cases/012.assets_archive/index.js
@@ -8,7 +8,15 @@ class MyResource extends pulumi.CustomResource {
             "asset": new pulumi.asset.StringAsset("foo"),
             "archive": new pulumi.asset.AssetArchive({}),
         });
-        super("test:index:MyResource", name, { "archive": archive });
+        let archiveP = Promise.resolve(new pulumi.asset.AssetArchive({
+            "foo": new pulumi.asset.StringAsset("bar"),
+        }));
+        let assetP = Promise.resolve(new pulumi.asset.StringAsset("baz"));
+        super("test:index:MyResource", name, {
+            "archive": archive,
+            "archiveP": archiveP,
+            "assetP": assetP,
+        });
     }
 }
 

--- a/sdk/nodejs/tests/runtime/langhost/run.spec.ts
+++ b/sdk/nodejs/tests/runtime/langhost/run.spec.ts
@@ -356,6 +356,40 @@ describe("rpc", () => {
             program: path.join(base, "012.assets_archive"),
             expectResourceCount: 1,
             registerResource: (ctx: any, dryrun: boolean, t: string, name: string, res: any) => {
+                assert.deepEqual(res, {
+                    "archive": {
+                        "4dabf18193072939515e22adb298388d": "0def7320c3a5731c473e5ecbe6d01bc7",
+                        "__pulumiArchive": true,
+                        "assets": {
+                            "archive": {
+                                "4dabf18193072939515e22adb298388d": "0def7320c3a5731c473e5ecbe6d01bc7",
+                                "__pulumiArchive": true,
+                                "assets": {},
+                            },
+                            "asset": {
+                                "4dabf18193072939515e22adb298388d": "c44067f5952c0a294b673a41bacd8c17",
+                                "__pulumiAsset": true,
+                                "text": "foo",
+                            },
+                        },
+                    },
+                    "archiveP": {
+                        "4dabf18193072939515e22adb298388d": "0def7320c3a5731c473e5ecbe6d01bc7",
+                        "__pulumiArchive": true,
+                        "assets": {
+                            "foo": {
+                                "4dabf18193072939515e22adb298388d": "c44067f5952c0a294b673a41bacd8c17",
+                                "__pulumiAsset": true,
+                                "text": "bar",
+                            },
+                        },
+                    },
+                    "assetP": {
+                        "4dabf18193072939515e22adb298388d": "c44067f5952c0a294b673a41bacd8c17",
+                        "__pulumiAsset": true,
+                        "text": "baz",
+                    },
+                });
                 return { urn: makeUrn(t, name), id: undefined, props: res };
             },
         },

